### PR TITLE
Support showing opcodes in the disassembly

### PR DESCRIPTION
**Pull request by lebr0nli**
This PR is trying to implement the feature that #1422 mentioned.
By setting `nearpc-num-opcode-bytes`(default is 0) and `nearpc-opcode-seperator-bytes` (default is 1):
- `set nearpc-num-opcode-bytes 9`
截圖 2023-01-24 下午8 25 49

- `set nearpc-num-opcode-bytes 4`

截圖 2023-01-24 下午8 26 34

- `set nearpc-num-opcode-bytes 9` + `set nearpc-opcode-seperator-bytes 0`
 截圖 2023-01-24 下午8 36 51

### CODE REVIEW

File: pwndbg/gdblib/nearpc.py [hunk 1]
```diff
@@ -162,7 +165,19 @@ def nearpc(pc=None, lines=None, emulate=False, repeat=False) -> List[str]:
...
+            opcodes = instr.bytes.hex()
+            opcodes = opcodes[: show_opcodes_size * 2]
+            align = show_opcodes_size * 2 + 10
+            if len(instr.bytes) > show_opcodes_size:
+                opcodes += pwndbg.color.gray("...")
+                align += 9  # len(pwndbg.color.gray(""))
+            opcodes = opcodes.ljust(align, " ")
+            if pwndbg.gdblib.config.highlight_pc and instr.address == pwndbg.gdblib.regs.pc:
+                opcodes = C.highlight(opcodes)
+            line = " ".join((prefix, address_str, opcodes, symbol, asm))
```
Comment (gsingh93) [inline pwndbg/gdblib/nearpc.py]: nit: Instead of duplicating this line here and in the `else` clause, we can probably add a `opcodes = ""` above the if statement and then move this right after the if statement, so both code paths use the same code to construct the output.

File: pwndbg/gdblib/nearpc.py [hunk 2]
```diff
@@ -162,7 +165,19 @@ def nearpc(pc=None, lines=None, emulate=False, repeat=False) -> List[str]:
...
             first_pc = False
 
-        line = " ".join((prefix, address_str, symbol, asm))
+        if show_opcodes_size > 0:
+            opcodes = instr.bytes.hex()
+            opcodes = opcodes[: show_opcodes_size * 2]
+            align = show_opcodes_size * 2 + 10
+            if len(instr.bytes) > show_opcodes_size:
+                opcodes += pwndbg.color.gray("...")
+                align += 9  # len(pwndbg.color.gray(""))
```
Comment (gsingh93) [inline pwndbg/gdblib/nearpc.py]: Can you add some comments explaining how the alignment is calculated? So if we show 9 opcode bytes (18 characters) we'll pad the end of it by 10 spaces, but if we add a '...' we pad the end by 19 spaces?

File: pwndbg/gdblib/nearpc.py [hunk 3]
```diff
@@ -54,7 +54,10 @@ def ljust_padding(lst):
     "nearpc-lines", 10, "number of additional lines to print for the nearpc command"
 )
 show_args = pwndbg.gdblib.config.add_param(
-    "nearpc-show-args", True, "show call arguments below instruction"
+    "nearpc-show-args", True, "whether to show call arguments below instruction"
+)
+show_opcodes_size = pwndbg.gdblib.config.add_param(
+    "nearpc-opcodes-size", 0, "number of bytes of opcodes to print for each instruction"
```
Comment (gsingh93) [inline pwndbg/gdblib/nearpc.py]: nit: "number of opcode bytes to print for each instruction"

Comment (gsingh93) [reply pwndbg/gdblib/nearpc.py]: And maybe the option name should be `nearpc-opcode-bytes`? Not sure.

**Review by gsingh93 — COMMENTED** [review]
This is great, thanks. Could you update the `nearpc` test to also test when `nearpc-opcode-size` is non-zero?

This doesn't need to be done in this PR, but I think having an option for adding spaces between each opcode byte would be nice.

File: pwndbg/gdblib/nearpc.py (hunk 3 above)
Comment (disconnect3d) [reply pwndbg/gdblib/nearpc.py]: or: `nearpc-show-opcode-bytes`?

**Comment by lebr0nli** [comment]
> Could you update the nearpc test to also test when nearpc-opcode-size is non-zero?

Sure.

> I think having an option for adding spaces between each opcode byte would be nice

Yeah, after some testing, seems like the readability with a space is better.
I make it default to one space for now and use `nearpc-opcode-seperator-bytes` for controlling it, but open to other suggestions.

File: pwndbg/gdblib/nearpc.py (hunk 3 above)
Comment (lebr0nli) [reply pwndbg/gdblib/nearpc.py]: Ok, I'll use `nearpc-show-opcode-bytes`

File: pwndbg/gdblib/nearpc.py (hunk 2 above)
Comment (lebr0nli) [reply pwndbg/gdblib/nearpc.py]: > Can you add some comments explaining how the alignment is calculated?

Sure.

> So if we show 9 opcode bytes (18 characters) we'll pad the end of it by 10 spaces, but if we add a '...' we pad the end by 19 spaces?

This 9 byte is because the gray color uses 9 invisible characters, so we need to extend the alignment for it to make it look the same.

File: pwndbg/gdblib/nearpc.py (hunk 1 above)
Comment (lebr0nli) [reply pwndbg/gdblib/nearpc.py]: Sure, I'll remove this `else` clause.

File: pwndbg/gdblib/nearpc.py (hunk 3 above)
Comment (gsingh93) [reply pwndbg/gdblib/nearpc.py]: `nearpc-show-opcode-bytes` makes it sound like a boolean. `nearpc-num-opcode-bytes` is another alternative.

Comment (lebr0nli) [reply pwndbg/gdblib/nearpc.py]: > `nearpc-show-opcode-bytes` makes it sound like a boolean. `nearpc-num-opcode-bytes` is another alternative.

Ok, I used `nearpc-num-opcode-bytes` now.

File: pwndbg/gdblib/config.py [hunk 4]
```diff
@@ -66,6 +79,15 @@ def get_set_string(self) -> str:
             # Anyway, to avoid some unexpected behaviors, we'll still set `self.param.value` to 0 here.
             self.param.value = 0
         else:
+            if self._custom_param_class:
+                if (self._custom_param_class == "PARAM_ZUINTEGER" and self.value < 0) or (  # type: ignore
+                    self._custom_param_class == "PARAM_ZUINTEGER_UNLIMITED" and self.value < -1  # type: ignore
```
Comment (disconnect3d) [inline pwndbg/gdblib/config.py]: Since this is making more and more complex and less and less clear, shall we maybe do something along the lines of:

```py
def __get_set_string_gdb_le_9(self): ...
def __get_set_string_gdb_gte_9(self): ...

if gdb_version >= 9:
    get_set_string = __get_set_string_gdb_gte_9
else:
    get_set_string = __get_set_string_gdb_le_9
```
@gsingh93 WDYT?

Comment (gsingh93) [reply pwndbg/gdblib/config.py]: +1, I think that'll be less complicated and cleaner.

**Review by disconnect3d — APPROVED** [review]
@gsingh93 I think we can merge this now?

**Comment by gsingh93** [comment]
Yup, lgtm.
